Add defined errors to the package

gdholtslander opened the pull request:
Hello! I have finally come back for #29 and solved it basically the way you suggested. I'm looking to try to get involved with more Go OSS and this is my first PR so far.

My biggest concern is that it's theoretically a breaking change, as if someone was depending on having error details in Error previously, those are now gone.

The reason this is so many changes is because as part of adding the ErrorDetails to Root I also changed the usages of Root to use labelled properties rather than specifying everything in each initialization.

Let me know if I should add an example to either the Readme or to the examples folder of this in use?

anaskhan96 commented:
Hi @gdholtslander 

First of all, thank you for the PR. This is amazing! Breaking changes are okay, as this will be a separate release.

I do have an issue though.
* Not sure if we should be including two error related fields in the `Root` struct. That could lead to two possible `nil pointer dereference` exceptions by a programmer instead of one on every success.
* On top of that, there's actually a duplication of error explanation text taking place. Let's take `ErrElementNotFound` for example. It has two strings, on in `Root.Error.Error()` which is "element not found", and one in `Root.ErrorDetails` which is "element x with attributes y not found"

Either we dispose of the `ErrorDetails`, or if we really want to give an explanation from our side, we have `Root.Error` be a variable of our custom composite type of error, something like this
```go
type SoupError struct {
    Type ErrorType // our enum with values `ErrUnableToParse`, `ErrElementNotFound`, etc.
    error
}
```
This way, we can get the type using `Root.Error.Type` and the detailed text using `Root.Error.Error()`

Let me know what you think.

Oh, and also unit test cases. Very important for an integral change like this :P

gdholtslander commented:
I like the second approach! I had that one essentially coded up but I was worried about backwards compatibility - essentially that people would already be checking for certain error strings and those wouldn't be there anymore. If it's still 1.x then there shouldn't be breaking changes... unless you meant we'd release it as a 2.0 if we added this. A bit extreme just for errors! :joy:

I think if we implement Error() as the old message, it should still be backward compatible...

I'll keep working on this. Thanks for the feedback!

gdholtslander commented:
I didn't realize it has been almost a year since I made the issue. Whoops 😨

gdholtslander commented:
Second approach we talked about is implemented, and I added some tests. I also added an example of some error handling one could do, but it's quite basic.

Looking forward to hearing what you think about this now.

gdholtslander commented on CHANGELOG.md:
```diff
@@ -1,4 +1,15 @@
-## v1.1
+## v1.2.0
```
I figured if we're versioning, we might as well go full [semver](https://semver.org/) here.

anaskhan96 replied:
Yes

gdholtslander commented on soup.go:
```diff
@@ -15,20 +15,67 @@ import (
+// from this package to see what kind of error it is, then further inspect the Error() method
+// to see if it has more specific details for you, like in the case of a ErrElementNotFound
+// type of error.
+type Error struct {
```
I did have this as SoupError initially, but my editor reminded me that this type will be used as `soup.SoupError` which means it will stutter. `soup.Error` looks decent, like in the example code.

anaskhan96 replied:
Yes let's keep it as `Error` alone

gdholtslander commented on soup.go:
```diff
@@ -72,24 +120,22 @@ func GetWithClient(url string, client *http.Client) (string, error) {
 // Get returns the HTML returned by the url in string using the default HTTP client
 func Get(url string) (string, error) {
-	// Init a new HTTP client
-	client := &http.Client{}
```
We can initialize this earlier and save some resources if someone is using repeated calls to `soup.Get`

anaskhan96 replied:
Sure.

anaskhan96 commented on examples/errors/errors.go:
```diff
@@ -0,0 +1,33 @@
+	}
+	xkcdSoup := soup.HTMLParse(xkcd)
+	links := xkcdSoup.Find("div", "id", "linkz")
+	if links.Error != nil && links.Error.(soup.Error).Type == soup.ErrElementNotFound {
```
@gdholtslander not such a huge fan of the type casting here
Why can't we just have `Root.Error` be a field of our custom error type?
That way, this line can be simplified to
```go
if links.Error != nil && links.Error.Type == soup.ErrElementNotFound {
```

anaskhan96 replied:
We're implementing the `error` interface and overriding with our own `Error()` method, which is right, i'm just thinking if it has any benefits

gdholtslander replied:
The biggest benefit is that it supports backwards compatibility. If you don't need the types you can continue exactly the way you were before by checking if the error is nil and doing your string checking if you wish (i.e. what I was doing, `strings.Contains(root.Error, "element not found")`). But once you're ready you can start using the types to check for certain errors instead.

I did make the realization while I was doing this that the actual errors we return from each method are always the same. For example, `Find` doesn't return `ElementNotFound` in some cases and `NextSiblingNotFound` in other cases. So there isn't a _ton_ of value in this PR, now that I've made this realization 🤦‍♂ But it was fun to do - and perhaps it is still useful to remove string checking from a project. And perhaps if we change how we compose some of our functions in the future, it would become more useful.

anaskhan96 replied:
Ah, the backward compatibility makes sense.

Yes this will certainly be useful when the library grows 👍🏻

anaskhan96 commented on soup.go:
```diff
@@ -15,20 +15,67 @@ import (
+// type of error.
+type Error struct {
+	Type ErrorType
+	msg  string
```
Since golang doesn't support inheritance but allows composition, i thought we could have it like this
```go
type Error struct {
    Type ErrorType
    error
}
```
and access it like `Error.Type` and `Error.Error()` (the public method exposed by the golang interface). But i just tried this on golang playground, and it looks like we can compose our struct with other __structs__, but not __interfaces__ (`error` is an interface) :/

gdholtslander replied:
IMO the way it actually works is even better than this, as what makes something an error is just that it has a method with a signature like `Error() string`. It's even more powerful than embedding an error 👍

anaskhan96 commented:
Merging this, thank you again for the PR! :)
